### ISSUE TITLE
Compiler ignores object type if any attribute has dynamic name

### ISSUE THREAD

**Issue by mgr01** [google/closure-compiler#2389]
See code below.
`anObject.shoudValidate()` has a non-null parameter.
It's called in `foobar()` with an optional one, but the compiler doesn't complain.

If you comment the dynamic attribute `['someOtherName']`, then the compilation fails.

```javascript
// ==ClosureCompiler==
// @output_file_name default.js
// @compilation_level ADVANCED_OPTIMIZATIONS
// ==/ClosureCompiler==

const anObject = {
  /** @param {string} requiredValue */
  shouldValidate: function(requiredValue) {},

  ['someOtherName']: function() {},
};

/** @param {?string} optionalValue */
function foobar(optionalValue) {
  anObject.shouldValidate(optionalValue);
}
```
